oauth_dropins/indieauth.py

```diff
@@ -0,0 +1,128 @@
+"""IndieAuth drop-in.
+
+https://indieauth.com/developers
+"""
+
+import json
+import logging
+import urllib
+import urlparse
+
+import appengine_config
+import handlers
+import models
+from webutil import util
+
+import bs4
+import requests
+from google.appengine.ext import ndb
+
+
+INDIEAUTH_URL = 'https://indieauth.com/auth'
+
+
+def discover_authorization_endpoint(me):
+  """Fetch a URL and look for authorization_endpoint Link header or
+  rel-value.
+
+  Args:
+    me: string, URL to fetch
+
+  Return:
+    string, the discovered indieauth URL or the default indieauth.com URL
+  """
+  logging.debug('trying to fetch user url "%s"', me)
+  resp = util.requests_get(me)
+  if resp.status_code < 200 or resp.status_code >= 300:
+    logging.debug(
+      'could not fetch user url "%s". got response code: %d',
+      me, resp.status_code)
+    return INDIEAUTH_URL
+  # check authorization_endpoint header first
+  auth_endpoint = resp.links.get('authorization_endpoint', {}).get('url')
+  if auth_endpoint:
+    return auth_endpoint
+  # check the html content
+  soup = bs4.BeautifulSoup(resp.text)
+  auth_link = soup.find('link', {'rel': 'authorization_endpoint'})
+  auth_endpoint = auth_link and auth_link['href']
+  if auth_endpoint:
+    return auth_endpoint
+  return INDIEAUTH_URL
+
+
+
+class IndieAuth(models.BaseAuth):
+  """An authenticated IndieAuth user.
+
+  Provides methods that return information about this user. Stores credentials
+  in the datastore. Key is the domain name. See models.BaseAuth for usage
+  details.
+  """
+  # access token
+  token = ndb.StringProperty(required=True)
+  user_json = ndb.TextProperty(required=True)  # generally this has only 'me'
+
+  def site_name(self):
+    return 'IndieAuth'
+
+  def user_display_name(self):
+    """Returns the user's domain."""
+    return self.key.string_id()
+
+  def access_token(self):
+    """Returns theAuth access token string."""
+    return self.token
+
+
+class StartHandler(handlers.StartHandler):
+  """Starts the IndieAuth flow. Requires the 'me' parameter with the
+  user URL that we want to authenticate.
+  """
+  def redirect_url(self, state=None):
+    assert appengine_config.INDIEAUTH_CLIENT_ID, (
+      "Please fill in the indieauth_client_id in your app's root directory.")
+
+    me = self.request.get('me')
+    redirect_uri = self.to_url()
+    endpoint = discover_authorization_endpoint(me)
+
+    url = endpoint + '?' + urllib.urlencode({
+      'me': me,
+      'client_id': appengine_config.INDIEAUTH_CLIENT_ID,
+      'redirect_uri': redirect_uri,
+      'state': state,
+    })
+
+    logging.info('Redirecting to IndieAuth: %s', url)
+    return str(url)
+
+
+class CallbackHandler(handlers.CallbackHandler):
+  """The callback handler from the IndieAuth request. POSTs back to the
+  auth endpoint to verify the authentication code."""
+  def get(self):
+    me = self.request.get('me')
+    code = self.request.get('code')
+    state = self.request.get('state')
```

Comment:
is this always provided? otherwise i worry that it will be `None` and `requests.post` below will `str` it to `'None'`...

maybe `self.request.get('state', '')`